spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/QueryMapper.java

```diff
@@ -455,6 +455,15 @@ public <T> T getPropertyValue(MongoPersistentProperty property) {
 				return collection.stream().map(it -> valueConverter.write(it, conversionContext)).collect(Collectors.toList());
 			}
 
+			if(!documentField.getProperty().isMap() && sourceValue instanceof Document document) {
+				return new Document(document.entrySet().stream().collect(Collectors.toMap(entry -> entry.getKey(), entry -> {
```

Comment:
minor suggestion: `.collect(Collectors.toMap(Entry::getKey, entry -> {`

Reply:
We had a second look and using the stream API here bears some risks. Entry order may not be preserved and the usage of Collectors API can cause NPE due to [JDK-8148463](https://bugs.openjdk.org/browse/JDK-8148463).